Added integration testing workflow for agent

bhautikpip opened the pull request:
*Issue #, if available:*

*Description of changes:*

1. Introducing integration test workflow for java agent
2. Added validator module to run validation to check different instrumentation (currently checks aws sdk, http (using okhttp) instrumentation)
3. Use Github's runner for set up collector and testing

Action items before merging:
1. Add aws account creds as secrets to agent repo
2. Reference validator module to master branch
3. Change this workflow to run on pull request only

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

bhautikpip commented:
Reference that workflow executes successfully: https://github.com/bhautikpip/aws-otel-java-instrumentation/runs/1288692867

anuraaga commented:
@bhautikpip Is this ready for review? Looks like the test still fails

bhautikpip commented:
it's not ready. I am still working on it (just small toke issues). However, feel free to review because workflow won't change drastically.

NathanielRN commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,113 @@
+        uses: actions/checkout@v2
+        with:
+          repository: aws-observability/aws-otel-test-framework-sample-apps
+          token: ${{  secrets.PAT }}
```
You'll probably have to get this `PAT` on the base repo (not your fork) if you want this to work because of some Github Actions [security policies](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull-request-events-for-forked-repositories) which sends the workflow to the base repo

bhautikpip replied:
Yes but I am trying to avoid having my PAT in org wide repo. Now, we are adding sample apps to SDK repo so we don't have to import this repo.

anuraaga commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,194 @@
+
+      - uses: actions/setup-java@v1
+        with:
+          java-version: 14
```
```suggestion
          java-version: 15
```

Guess may as well use latest (unless it actually doesn't build for some reason)

anuraaga commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,194 @@
+      - name: Run Spring Boot App with AWS Java agent
+        run: |
+          agent_jar_name=`ls aws-opentelemetry-agent-*`
+          java -javaagent:$agent_jar_name -jar springboot-1.0.jar &
```
I'm a bit surprised we're running directly without docker. But I guess it's ok if it works

bhautikpip replied:
I am planning to add docker. This is just and initial draft where I wanted to see sample apps are up and running as expected.

anuraaga commented on sample-apps/spark/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM gradle:6.5.0-jdk11 as builder
```
Though then I noticed we have this Dockerfile - do we need the dockerfiles?

bhautikpip replied:
We do need dockerfiles for sample apps. Since we are planning to build docker images and publish to docker hubs for other repos to consume.

anuraaga commented on sample-apps/spark/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM gradle:6.5.0-jdk11 as builder
+
+# build
```
If using docker, recommend using jib instead of docker, can just use this base image which has the agent preloaded.

https://github.com/aws-observability/aws-otel-java-instrumentation/blob/main/otelagent/build.gradle.kts#L83

anuraaga commented on sample-apps/spark/build.gradle:
```diff
@@ -0,0 +1,53 @@
+    extraLibs
+}
+
+jar {
```
If you need a fatjar I recommend spring-boot plugin. Even when not using spring boot it works great for making a fatjar.

Though if we go with docker, we can just not generate fatjar, I don't think we need to support both patterns.

bhautikpip replied:
Okay we will go with the docker approach you suggested.

anuraaga commented on sample-apps/spark/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.2.1-bin.zip
```
Ancient gradle version! Let's not have subgradle wrappers for the sample apps since they're not necessarily for user consumption, e.g., copy-paste and use. So just make them subprojects of the top project.

bhautikpip replied:
good catch!

anuraaga commented on sample-apps/spark/src/main/java/com/amazon/sampleapp/App.java:
```diff
@@ -0,0 +1,131 @@
+/*
```
Remove license headers from sample apps and add a LICENSE file in the sample-apps folder saying these are MIT-0 licensed because they are sample apps

anuraaga replied:
I guess we'll need to tweak the spotless setting in the `sample-apps` project to fix the license. Can be followup though

anuraaga reviewed (commented):
Left some high level comments, especially removing the sub-gradle wrappers will probably change things quite a bit (apply formatting, etc)

bhautikpip commented:
Hi @anuraaga ,

So I addressed most of your comments. Feel free to review it. Currently, this workflow publishes sample app builds (with released version of agent) to ECR. Since, for collector testing we would want the sample app image with latest released java agent version and not the developing version (master branch). For integration testing workflow we are testing against developing version of an agent.

anuraaga reviewed (commented):
I think we may as well publish the master version of the agent with this workflow to the docker registry (and probably drop the fatjar). A similar workflow would also run with release-build.yml, and that can publish the release version, which collector or whoever can consume.

bhautikpip commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,147 @@
+        with:
+          arguments: build jibDockerBuild --stacktrace -PenableCoverage=true
+        env:
+          AWS_REGISTERY_ACCOUNT: ${{ secrets.AWS_ACCOUNT }}
```
Done

bhautikpip commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,147 @@
+      - name: Build with Gradle
+        uses: burrunan/gradle-cache-action@v1
+        with:
+          arguments: build jibDockerBuild --stacktrace -PenableCoverage=true
```
Okay.

bhautikpip commented on .github/workflows/integration-testing.yml:
```diff
@@ -0,0 +1,147 @@
+
+      - name: Run Spark App with Java agent
+        run: |
+          docker run -e AWS_ACCESS_KEY_ID=${{ secrets.AWS_ACCESS_KEY_ID }} -e AWS_SECRET_ACCESS_KEY=${{ secrets.AWS_SECRET_ACCESS_KEY }} -e AWS_REGION=us-west-2 -e INSTANCE_ID=${{ github.run_id }}-${{ github.run_number }} -e LISTEN_ADDRESS=0.0.0.0:4567 -e OTEL_EXPORTER_OTLP_ENDPOINT=172.17.0.1:55680 -p 4567:4567 ${{ steps.login-ecr.outputs.registry }}/spark:${{ github.sha }} &
```
Yes, we do need aws creds as within sample apps we are making an aws call (call to s3).

wyTrivail replied:
if the s3 call fails, will the trace still be generated?

bhautikpip replied:
Yes. I think so!

anuraaga commented:
I added https://github.com/aws-observability/aws-otel-java-instrumentation/pull/6/commits/2de27275d34208a1a44113a05e1cca9a534fdefd to cleanup the tooling, in particular checking that header exists on files that aren't sample-apps, and remove headers from sample app files since MIT-0 shouldn't have headers.

anuraaga commented on sample-apps/springboot/src/main/java/com/amazon/sampleapp/DemoApplication.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    Map<String, Object> config =
+        new HashMap<String, Object>() {
```
Just call `put` normally instead of making an anonymous class, this pattern is discouraged

bhautikpip replied:
Sure.

anuraaga commented on sample-apps/springboot/src/main/java/com/amazon/sampleapp/DemoApplicationInterceptor.java:
```diff
@@ -0,0 +1,56 @@
+        metricEmitter.emitBytesSentMetric(
+            request.getContentLength() + mimicPayloadSize(), request.getServletPath(), statusCode);
+      } catch (Exception e) {
+        e.printStackTrace();
```
Is this the right way to handle this error?

bhautikpip replied:
think so because we would be getting general emitter related errors inside try block.

anuraaga commented on sample-apps/spark/src/main/java/com/amazon/sampleapp/MetricEmitter.java:
```diff
@@ -0,0 +1,75 @@
+import io.opentelemetry.metrics.LongValueRecorder;
+import io.opentelemetry.metrics.Meter;
+
+public class MetricEmitter {
```
What's the point of `MetricEmitter`? It seems to be the same in both sample apps, so it's not actually related to a framework

bhautikpip replied:
MetricEmitter is related to a sample app. We are performing validation on Metric Collected by SDK and sample app only emits metric on `/outgoing-http-call` path. MetricEmitter is required for every sample apps.

anuraaga replied:
I think I get it - but it is copy-paste code among all the apps. I guess it'd be good if we can publish a testing-support library to maven central eventually.
